src/actions.js

```diff
@@ -350,52 +366,84 @@ export const startWallet = (words, pin) => async (dispatch) => {
     }
   });
 
+  const handlePartialUpdate = async (updatedBalanceMap) => {
+    const tokens = Object.keys(updatedBalanceMap);
+    const tokensHistory = {};
+    const tokensBalance = {};
+
+    for (const token of tokens) {
+      /* eslint-disable no-await-in-loop */
+      const balance = await wallet.getBalance(token);
+      const tokenBalance = balance[0].balance;
+      tokensBalance[token] = {
+        available: tokenBalance.unlocked,
+        locked: tokenBalance.locked,
+      };
+      const history = await wallet.getTxHistory({ token_id: token });
+      tokensHistory[token] = history.map((element) => mapTokenHistory(element, token));
+      /* eslint-enable no-await-in-loop */
+    }
+
+    dispatch(partiallyUpdateHistoryAndBalance({ tokensHistory, tokensBalance }));
+  };
+
   wallet.start({ pinCode: pin, password: pin }).then((serverInfo) => {
     walletUtil.storePasswordHash(pin);
     walletUtil.storeEncryptedWords(words, pin);
+
     dispatch(setServerInfo({ version: null, network: networkName }));
 
-    if (!useWalletService) {
-      wallet.on('new-tx', (tx) => {
-        fetchNewTxTokenBalance(wallet, tx).then((updatedBalanceMap) => {
-          if (updatedBalanceMap) {
-            dispatch(newTx(tx, updatedBalanceMap));
-          }
-        });
-      });
+    wallet.on('new-tx', (tx) => {
+      fetchNewTxTokenBalance(wallet, tx).then(async (updatedBalanceMap) => {
+        if (updatedBalanceMap) {
+          dispatch(newTx(tx, updatedBalanceMap));
 
-      wallet.on('update-tx', (tx) => {
-        fetchNewTxTokenBalance(wallet, tx).then((updatedBalanceMap) => {
-          if (updatedBalanceMap) {
-            dispatch(updateTx(tx, updatedBalanceMap));
+          // Since we are not keeping track of the tokens balance and history
+          // on the wallet service facade, we need to fetch it from the server
+          // on every new transaction received
+          if (useWalletService) {
+            handlePartialUpdate(updatedBalanceMap);
```

Comment:
This methods calculates the balance of each token just like `fetchNewTxTokenBalance` and you update the redux on `dispatch(newTx)` above. Why do you need it?

Reply:
You are right, this is a bug actually...

The `onNewTx` reducer will try to update the balances using the `getTxBalance` method from the old facade that is synchronous, but the `getTxBalance` from the wallet service is asyncronous. I'm keeping the `onNewTx` call because it is needed for the request payment feature, but returning before the balance calculation in case we are using the wallet service facade, and using the new `handlePartialUpdate` method for it.

Done in 53e2975ac7c52a0c6195969184122ff8b5f823b

Reply:
Done in 53e2975ac7c52a0c6195969184122ff8b5f823b

Reply:
Why don't you change `getTxBalance` method to be async so you don't need to do specific code for both facades? Both facades are supposed to implement the same interface, so we don't need to do things so differently.

Reply:
That won't work because reducers need to be pure functions, so they can't be async

`getTxBalance` currently needs to be async on the Wallet Service because we fetch the wallet's addresses from the Wallet Service every time we call it.

We could refactor it to keep the addresses in memory and keeping it updated somehow, this way `getTxBalance` could work without being async.

That would be the only way to maintain compatibility for both facades on the reducer

What do you think?